src/Microsoft.OData.Edm/EdmModelVisitor.cs

```diff
@@ -299,6 +433,35 @@ public void VisitTypeReference(IEdmTypeReference reference)
             }
         }
 
+        public Task VisitTypeReferenceAsync(IEdmTypeReference reference)
+        {
+            switch (reference.TypeKind())
+            {
+                case EdmTypeKind.Collection:
+                    return this.ProcessCollectionTypeReferenceAsync(reference.AsCollection());
+                case EdmTypeKind.Complex:
+                    return this.ProcessComplexTypeReferenceAsync(reference.AsComplex());
+                case EdmTypeKind.Entity:
+                    return this.ProcessEntityTypeReferenceAsync(reference.AsEntity());
+                case EdmTypeKind.EntityReference:
+                    return this.ProcessEntityReferenceTypeReferenceAsync(reference.AsEntityReference());
+                case EdmTypeKind.Enum:
+                    return this.ProcessEnumTypeReferenceAsync(reference.AsEnum());
+                case EdmTypeKind.Primitive:
+                    return this.VisitPrimitiveTypeReferenceAsync(reference.AsPrimitive());
+                case EdmTypeKind.TypeDefinition:
+                    return this.ProcessTypeDefinitionReferenceAsync(reference.AsTypeDefinition());
+                case EdmTypeKind.None:
+                    return this.ProcessTypeReferenceAsync(reference);
+                case EdmTypeKind.Path:
+                    return this.ProcessPathTypeReferenceAsync(reference.AsPath());
+                case EdmTypeKind.Untyped:
+                    return this.ProcessUntypedTypeReferenceAsync(reference as IEdmUntypedTypeReference);
+                default:
+                    throw new InvalidOperationException(Edm.Strings.UnknownEnumVal_TypeKind(reference.TypeKind().ToString()));

```

Comment:
See my reasoning elsewhere https://github.com/OData/odata.net/pull/3081/files#r1796808438

Reply:
Also looking at the `switch` statement and the cases, the `default` code block will not be reachable and as a result just to be careful, I will return a default task.

Not possible to write a test for this at this time because all the `EdmTypeKind` have been checked.

Reply:
@gathogojr same as above. I have opted to use Contract.Assert and return faulted task:

...
 Contract.Assert(false, Edm.Strings.UnknownEnumVal_ExpressionKind(expression.ExpressionKind));
 return Task.FromException<InvalidOperationException>(new InvalidOperationException(Edm.Strings.UnknownEnumVal_ExpressionKind(expression.ExpressionKind)));
...